### DISCUSSION TITLE
Composable Archtitecture Tests: Data pointer equality check necessary?

### DISCUSSION THREAD

**Discussion by mschonvogel** [pointfreeco/swift-composable-architecture#628] [Q&A]
I am currently testing a store and the fact that the pointer of a Data Object in my state is being compared makes it impossible for me to test that part with a mock. Is there a way to work around this?

For two data objects to be equal it should be enough if they hold the same number of bytes, and if the bytes at the same position in the objects are the same.

```
...            
value: Data(
           count: 414619,
    −      pointer: 0x0000000111d63000
    +      pointer: 0x0000000111cfd000
)
...
```

**Comment by iampatbrown** [comment]
Can you provide a little more info? Are you testing the store via TestStore or just trying to get it working? 
If it's a ViewStore you can initialise it with a custom isDuplicate check which could ignore the pointer.

**Reply by mschonvogel** [reply]
Sure, here you go:

```swift
        let uuid = UUID()
        let createdAt = Date()
        let image = UIImage(named: "document", in: .module, with: nil)!
        let data = image.jpegData(compressionQuality: 1)
        
        let store = TestStore(
            initialState: DocumentListState(),
            reducer: documentListReducer,
            environment: .failing()
        )
        
        store.send(.imagePickerCompleted(result: .success([image]))) {
            $0.attachments = [
                Attachment(
                    id: uuid,
                    title: "Img_53A.jpg",
                    data: data,
                    createdAt: createdAt
                )
            ]
        }
```

**Reply by iampatbrown** [reply]
Are you able to add a custom equality check to DocumentListState or Attachment (is that your type?)? 

```swift
extension DocumentListState {
  static func == (lhs: Self, rhs: Self) -> Bool {
    // add your equality check here
  }
}
```

**Reply by iampatbrown** [reply]
How does the imagePickerCompleted action work?
Are you calling .jpegData(compressionQuality: 1) inside the reducer? Maybe you need to pull this into the environment and mock it there. Just not sure how the uuid and data are given to the attachment via the action atm based on the code above.
